src/lib.rs

```diff
@@ -108,19 +184,23 @@ mod wasm {
     pub(super) struct WASMValue(Option<Value>);
 
     impl WASMValue {
-        pub(super) fn value(self) -> Value {
-            self.0.unwrap()
+        pub(super) fn value(self) -> Result<Value, WasmError> {
+            self.0.ok_or(WasmError::NoValue)
         }
-
-        pub(super) fn i32(self) -> i32 {
-            i32::try_from(self.0.unwrap()).expect("expected an i32 value")
+        pub(super) fn i32(self) -> Result<i32, WasmError> {
+            self.0
+                .and_then(|val| val.i32())
+                .ok_or(WasmError::InvalidI32)
         }
-
-        pub(super) fn bool(self) -> bool {
+        pub(super) fn u32(self) -> Result<u32, WasmError> {
+            let value = self.i32()?;
+            u32::try_from(value).map_err(|source| WasmError::InvalidU32 { value, source })
+        }
+        pub(super) fn bool(self) -> Result<bool, WasmError> {
             match self.i32() {
-                0 => false,
-                1 => true,
-                _ => panic!("expected a boolean value"),
+                Ok(0) => Ok(false),
+                Ok(1) => Ok(true),
+                _ => Err(WasmError::InvalidBool),
             }
         }
```

Comment:
I wonder if these should be From impls 🤔

Reply:
Yeah, I don't have strong feelings either way on this. It's all internal to the crate so we can change this in a followup PR if necessary.

Reply:
I made them TryFrom and I like it a lot.